File: test/i386/srcs/thread_test.c
//
// This file is part of libdebug Python library (https://github.com/libdebug/libdebug).
// Copyright (c) 2024 Roberto Alessandro Bertolini. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for details.
//

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>

void thread_1_function()
{
    asm volatile (
    "mov $0x00112233, %%eax\n\t"
    "nop\n\t"::: "eax");
}

void thread_2_function()
{
    asm volatile (
    "mov $0xccdd1122, %%eax\n\t"
    "nop\n\t"::: "rax");
}

void thread_3_function()
{
    asm volatile (
    "mov $0x66770011, %%eax\n\t"
    "nop\n\t"::: "rax");
}

void do_nothing()
{
    asm volatile ("nop\n\t");
}

int main()
{
    pthread_t thread_1, thread_2, thread_3;
    pthread_create(&thread_1, NULL, (void *)thread_1_function, NULL);
    pthread_create(&thread_2, NULL, (void *)thread_2_function, NULL);
    pthread_create(&thread_3, NULL, (void *)thread_3_function, NULL);
    pthread_join(thread_1, NULL);
    pthread_join(thread_2, NULL);
    pthread_join(thread_3, NULL);

    do_nothing();

    return 0;
}

File: test/aarch64/scripts/thread_test.py
#
# This file is part of libdebug Python library (https://github.com/libdebug/libdebug).
# Copyright (c) 2024 Roberto Alessandro Bertolini, Gabriele Digregorio. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project root for details.
#

import unittest

from libdebug import debugger


class ThreadTest(unittest.TestCase):
    def setUp(self):
        pass

    def test_thread(self):
        d = debugger("binaries/thread_test")

        d.run()

        bp_t0 = d.breakpoint("do_nothing")
        bp_t1 = d.breakpoint("thread_1_function")
        bp_t2 = d.breakpoint("thread_2_function")
        bp_t3 = d.breakpoint("thread_3_function")

        t1_done, t2_done, t3_done = False, False, False

        d.cont()

        for _ in range(150):
            if bp_t0.address == d.regs.pc:
                self.assertTrue(t1_done)
                self.assertTrue(t2_done)
                self.assertTrue(t3_done)
                break

            if len(d.threads) > 1 and bp_t1.address == d.threads[1].regs.pc:
                t1_done = True
            if len(d.threads) > 2 and bp_t2.address == d.threads[2].regs.pc:
                t2_done = True
            if len(d.threads) > 3 and bp_t3.address == d.threads[3].regs.pc:
                t3_done = True

            d.cont()

        d.kill()
        d.terminate()

    def test_thread_hardware(self):
        d = debugger("binaries/thread_test")

        d.run()

        bp_t0 = d.breakpoint("do_nothing", hardware=True)
        bp_t1 = d.breakpoint("thread_1_function", hardware=True)
        bp_t2 = d.breakpoint("thread_2_function", hardware=True)
        bp_t3 = d.breakpoint("thread_3_function", hardware=True)

        t1_done, t2_done, t3_done = False, False, False

        d.cont()

        for _ in range(15):
            if bp_t0.address == d.regs.pc:
                self.assertTrue(t1_done)
                self.assertTrue(t2_done)
                self.assertTrue(t3_done)
                break

            if len(d.threads) > 1 and bp_t1.address == d.threads[1].regs.pc:
                t1_done = True
            if len(d.threads) > 2 and bp_t2.address == d.threads[2].regs.pc:
                t2_done = True
            if len(d.threads) > 3 and bp_t3.address == d.threads[3].regs.pc:
                t3_done = True

            d.cont()

        d.kill()
        d.terminate()

if __name__ == "__main__":
    unittest.main()
